oraclepaas/resource_mysql_service_instance.go

```diff
@@ -0,0 +1,662 @@
+		input.MeteringFrequency = value.(string)
+	}
+
+	if value, ok := d.GetOk("region"); ok {
+		input.Region = value.(string)
+	}
+
+	if value, ok := d.GetOk("service_description"); ok {
+		input.ServiceDescription = value.(string)
+	}
+
+	err := expandCloudStorage(d, input)
+	if err != nil {
+		return *input, err
+	}
+
+	return *input, nil
+}
+
+/**
+Expands and reads the values in the Cloud_Storage list specified in the terraform file.
+*/
+func expandCloudStorage(d *schema.ResourceData, parameter *mysql.ServiceParameters) error {
+
+	cloudStorageInfo := d.Get("cloud_storage_configuration").([]interface{})
+
+	if parameter.BackupDestination == string(mysql.ServiceInstanceBackupDestinationBoth) || parameter.BackupDestination == string(mysql.ServiceInstanceBackupDestinationOSS) {
+		if len(cloudStorageInfo) == 0 {
+			return fmt.Errorf("`cloud_storage_configuration` must be set if `backup_destination` is set to `OSS` or `BOTH`")
+		}
+	}
+
+	if len(cloudStorageInfo) > 0 {
+		attrs := cloudStorageInfo[0].(map[string]interface{})
+		parameter.CloudStorageContainer = attrs["cloud_storage_container"].(string)
+		parameter.CloudStorageContainerAutoGenerate = attrs["create_if_missing"].(bool)
+		if val, ok := attrs["cloud_storage_username"].(string); ok && val != "" {
+			parameter.CloudStorageUsername = val
+		}
+		if val, ok := attrs["cloud_storage_password"].(string); ok && val != "" {
+			parameter.CloudStoragePassword = val
+		}
+	}
+	return nil
+}
+
+func expandEM(input map[string]interface{}, parameter *mysql.MySQLParameters) error {
+
+	emInfo := input["enterprise_monitor_configuration"].([]interface{})
+
+	log.Printf("[DEBUG] input                       : %v", input)
+	log.Printf("[DEBUG] parameter.EnterpriseMonitor : %v", parameter.EnterpriseMonitor)
+	log.Printf("[DEBUG] emInfo                      : %d", len(emInfo))
+
+	if parameter.EnterpriseMonitor == "Yes" {
+		if len(emInfo) == 0 {
+			return fmt.Errorf("`enterprise_monitor_configuration` must be set if `enterprise_monitor` is set to `Yes`")
+		}
+	}
+
+	if len(emInfo) > 0 {
+		attrs := emInfo[0].(map[string]interface{})
+
+		if val, ok := attrs["em_agent_password"].(string); ok && val != "" {
+			parameter.EnterpriseMonitorAgentPassword = attrs["em_agent_password"].(string)
+		}
+
+		if val, ok := attrs["em_agent_user"].(string); ok && val != "" {
+			parameter.EnterpriseMonitorAgentUser = attrs["em_agent_user"].(string)
+		}
+
+		if val, ok := attrs["em_password"].(string); ok && val != "" {
+			parameter.EnterpriseMonitorManagerPassword = attrs["em_password"].(string)
+		}
+
+		if val, ok := attrs["em_user"].(string); ok && val != "" {
+			parameter.EnterpriseMonitorManagerUser = attrs["em_user"].(string)
+		}
+
+		if val, ok := attrs["em_port"].(string); ok && val != "" {
+			parameter.MysqlEMPort = attrs["em_port"].(string)
+		}
+	}
+
+	return nil
+}
+
+func getComponentParameters(d *schema.ResourceData) (mysql.ComponentParameters, error) {
```

Comment:
Let's refactor this to `expandComponentParameters`

Reply:
Done